compiler/rustc_codegen_ssa/src/back/symbol_export.rs

```diff
@@ -321,6 +321,39 @@ fn exported_symbols_provider_local(
 
         let cgus = tcx.collect_and_partition_mono_items(()).codegen_units;
 
+        // Do not export symbols that cannot be instantiated by downstream crates.
+        let reachable_set = tcx.reachable_set(());
+        let is_local_to_current_crate = |ty: Ty<'_>| {
+            let no_refs = ty.peel_refs();
+            let root_def_id = match no_refs.kind() {
+                rustc_middle::ty::Closure(closure, _) => *closure,
+                rustc_middle::ty::FnDef(def_id, _) => *def_id,
+                rustc_middle::ty::Coroutine(def_id, _) => *def_id,
+                rustc_middle::ty::CoroutineClosure(def_id, _) => *def_id,
+                rustc_middle::ty::CoroutineWitness(def_id, _) => *def_id,
+                rustc_middle::ty::Foreign(def_id) => *def_id,
```

Comment:
Why is `Foreign` included here but not `Adt`?

Reply:
`Adt` was initially included, but it led to false positives with drop glue. I believe we did not export drop glue of types that are not reachable, even though it might've been referenced by drop glue of other types. This led to extra instantiations downstream.

Initially this match simply included every variant with DefId, so I'm fine with not including `Foreign` here.

Reply:
I've removed Foreign from this match; we should be fine with the rest.